### ISSUE TITLE
Karpenter in unconsolidatable state when there are any unschedulable pods

### ISSUE THREAD

**Issue by rkilingr** [kubernetes-sigs/karpenter#678]
## **Observed Behavior**:
When there are any unschedulable pods sitting anywhere in the cluster, karpenter fails to consolidate the existing pods towards right sized instance types. This happens even when the unschedulable pods are unrelated(managed by CA, Node Group) and would never schedule on the karpenter provisioned node due to different `tolerations` and `nodeSelector`

## **Expected Behavior**:
Shouldn't karpenter ignore unrelated pods being unschedulable when consolidating pods of a particular provisioner?
### Example:
- Unschedulable pods:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: fake-sample-service-usw2-dev2
  labels:
    name: fake-sample-service-usw2-dev2
---
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: fake-sample-service-usw2-dev2-default
spec:
  limits:
    resources:
      cpu: 20
  consolidation:
    enabled: true
  taints:
    - key: ig/fakesampleserviceusw2dev2default
      effect: NoSchedule
    - key: ig/fakesampleserviceusw2dev2default
      effect: NoExecute
  labels:
    node.kubernetes.io/instancegroup: fake-sample-service-usw2-dev2-default
    node.kubernetes.io/role: fake-sample-service-usw2-dev2-default
  providerRef:
    name: custom-family
  requirements:
    - key: karpenter.sh/capacity-type
      operator: In
      values:
        - on-demand
    - key: karpenter.k8s.aws/instance-category
      operator: In
      values: ['m','c','r']
    - key: karpenter.k8s.aws/instance-generation
      operator: Gt
      values: ['5']
    - key: kubernetes.io/os
      operator: In
      values:
        - linux
    - key: kubernetes.io/arch
      operator: In
      values:
        - amd64
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: fake-sample-service
  name: fake-sample-service-deployment
  namespace: fake-sample-service-usw2-dev2
spec:
  replicas: 4
  selector:
    matchLabels:
      app: fake-sample-service
  template:
    metadata:
      labels:
        app: fake-sample-service
    spec:
      tolerations:
        - key: ig/fakesampleserviceusw2dev2default
      nodeSelector:
        node.kubernetes.io/instancegroup: fake-sample-service-usw2-dev2-default
      containers:
        - image: nginx:1.14.2
          name: app
          resources:  {"limits":{"cpu":"30","memory":"10Gi"},"requests":{"cpu":"30","memory":"10Gi"}}
```
- Pods that are supposed to consolidate:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: fake-sample-service-usw2-dev
  labels:
    name: fake-sample-service-usw2-dev
---
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: fake-sample-service-usw2-dev-default
spec:
  consolidation:
    enabled: true
  taints:
    - key: ig/fakesampleserviceusw2devdefault
      effect: NoSchedule
    - key: ig/fakesampleserviceusw2devdefault
      effect: NoExecute
  labels:
    node.kubernetes.io/instancegroup: fake-sample-service-usw2-dev-default
    node.kubernetes.io/role: fake-sample-service-usw2-dev-default
  providerRef:
    name: custom-family
  requirements:
    - key: karpenter.sh/capacity-type
      operator: In
      values:
        - on-demand
    - key: karpenter.k8s.aws/instance-category
      operator: In
      values: ['m','c','r']
    - key: karpenter.k8s.aws/instance-generation
      operator: Gt
      values: ['5']
    - key: kubernetes.io/os
      operator: In
      values:
        - linux
    - key: kubernetes.io/arch
      operator: In
      values:
        - amd64
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: fake-sample-service
  name: fake-sample-service-deployment
  namespace: fake-sample-service-usw2-dev
spec:
  replicas: 1
  selector:
    matchLabels:
      app: fake-sample-service
  template:
    metadata:
      labels:
        app: fake-sample-service
    spec:
      tolerations:
        - key: ig/fakesampleserviceusw2devdefault
      nodeSelector:
        node.kubernetes.io/instancegroup: fake-sample-service-usw2-dev-default
      containers:
        - image: nginx:1.14.2
          name: app
          resources:  {"limits":{"cpu":"2","memory":"4Gi"},"requests":{"cpu":"2","memory":"4Gi"}}
```

Observation is a `c6a.4xlarge` instance is created and then the nodes have the following event preventing it from consolidation:
```shell
Normal   Unconsolidatable         8m58s              karpenter        not all pods would schedule, fake-sample-service-usw2-dev2/fake-sample-service-deployment-7fb449d879-q95hj => incompatible with provisioner "custom-provideref", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, incompatible requirements, key node.kubernetes.io/instancegroup, node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default] not in node.kubernetes.io/instancegroup In [knodes]; incompatible with provisioner "fake-sample-service-usw2-dev2-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, no instance type satisfied resources {"cpu":"30490m","memory":"11675723008","pods":"10"} and requirements karpenter.k8s.aws/instance-category In [c m r], karpenter.k8s.aws/instance-generation Exists >5, karpenter.sh/capacity-type In [on-demand], karpenter.sh/provisioner-name In [fake-sample-service-usw2-dev2-default], kubernetes.io/arch In [amd64], kubernetes.io/os In [linux], node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default], node.kubernetes.io/role In [fake-sample-service-usw2-dev2-default] (no instance type has enough resources); incompatible with provisioner "fake-sample-service-usw2-dev-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, did not tolerate ig/fakesampleserviceusw2devdefault=:NoSchedule; did not tolerate ig/fakesampleserviceusw2devdefault=:NoExecute fake-sample-service-usw2-dev2/fake-sample-service-deployment-7fb449d879-qpcz7 => incompatible with provisioner "custom-provideref", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, incompatible requirements, key node.kubernetes.io/instancegroup, node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default] not in node.kubernetes.io/instancegroup In [knodes]; incompatible with provisioner "fake-sample-service-usw2-dev2-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, no instance type satisfied resources {"cpu":"30490m","memory":"11675723008","pods":"10"} and requirements karpenter.k8s.aws/instance-category In [c m r], karpenter.k8s.aws/instance-generation Exists >5, karpenter.sh/capacity-type In [on-demand], karpenter.sh/provisioner-name In [fake-sample-service-usw2-dev2-default], kubernetes.io/arch In [amd64], kubernetes.io/os In [linux], node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default], node.kubernetes.io/role In [fake-sample-service-usw2-dev2-default] (no instance type has enough resources); incompatible with provisioner "fake-sample-service-usw2-dev-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, did not tolerate ig/fakesampleserviceusw2devdefault=:NoSchedule; did not tolerate ig/fakesampleserviceusw2devdefault=:NoExecute fake-sample-service-usw2-dev2/fake-sample-service-deployment-7fb449d879-dvrp4 => incompatible with provisioner "custom-provideref", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, incompatible requirements, key node.kubernetes.io/instancegroup, node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default] not in node.kubernetes.io/instancegroup In [knodes]; incompatible with provisioner "fake-sample-service-usw2-dev2-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, no instance type satisfied resources {"cpu":"30490m","memory":"11675723008","pods":"10"} and requirements karpenter.k8s.aws/instance-category In [c m r], karpenter.k8s.aws/instance-generation Exists >5, karpenter.sh/capacity-type In [on-demand], karpenter.sh/provisioner-name In [fake-sample-service-usw2-dev2-default], kubernetes.io/arch In [amd64], kubernetes.io/os In [linux], node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default], node.kubernetes.io/role In [fake-sample-service-usw2-dev2-default] (no instance type has enough resources); incompatible with provisioner "fake-sample-service-usw2-dev-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, did not tolerate ig/fakesampleserviceusw2devdefault=:NoSchedule; did not tolerate ig/fakesampleserviceusw2devdefault=:NoExecute fake-sample-service-usw2-dev2/fake-sample-service-deployment-7fb449d879-nbpv6 => incompatible with provisioner "custom-provideref", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, incompatible requirements, key node.kubernetes.io/instancegroup, node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default] not in node.kubernetes.io/instancegroup In [knodes]; incompatible with provisioner "fake-sample-service-usw2-dev2-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, no instance type satisfied resources {"cpu":"30490m","memory":"11675723008","pods":"10"} and requirements karpenter.k8s.aws/instance-category In [c m r], karpenter.k8s.aws/instance-generation Exists >5, karpenter.sh/capacity-type In [on-demand], karpenter.sh/provisioner-name In [fake-sample-service-usw2-dev2-default], kubernetes.io/arch In [amd64], kubernetes.io/os In [linux], node.kubernetes.io/instancegroup In [fake-sample-service-usw2-dev2-default], node.kubernetes.io/role In [fake-sample-service-usw2-dev2-default] (no instance type has enough resources); incompatible with provisioner "fake-sample-service-usw2-dev-default", daemonset overhead={"cpu":"490m","memory":"938304768","pods":"9"}, did not tolerate ig/fakesampleserviceusw2devdefault=:NoSchedule; did not tolerate ig/fakesampleserviceusw2devdefault=:NoExecute
```

whereas without the unschedulable pods, it consolidates the node to `c6a.xlarge`
## **Versions**:
- Chart Version: v0.29.2
- Kubernetes Version (`kubectl version`): 
  ```shell
  Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.16", 
  GitCommit:"51e33fadff13065ae5518db94e84598293965939", GitTreeState:"clean", BuildDate:"2023-07-19T12:26:21Z", 
  GoVersion:"go1.20.6", Compiler:"gc", Platform:"darwin/amd64"}
  Kustomize Version: v4.5.4
  Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.16-eks-2d98532", 
  GitCommit:"af930c12e26ef9d1e8fac7e3532ff4bcc1b2b509", GitTreeState:"clean", BuildDate:"2023-07-28T16:52:47Z", 
  GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
  ```

## Notes
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by njtran** [comment]
@rkilingr, can you share the same information you did at the beginning of the issue to confirm if this is an issue or not? This should be fixed in v0.30.0

**Comment by rkilingr** [comment]
I've verified that the issue is resolved with current version(v0.34.0) with NodePool and EC2NodeClass CR's configured, 

Thanks!
